### ISSUE TITLE
Podcast behavior 

### ISSUE THREAD

**Issue by llucps** [airsonic/airsonic#1392]
Hi,

This is not really bug but how Podcasts work in Airsonic. 
The current workflow of listening a new episode, involves first download the episode and then play it, that's totally fine with me, it would be great if you had the option to just stream it but it's not a big deal.

The main problem is once I delete an episode that has been downloaded, it deletes not just the file but also the reference in the list of episodes, so once it's deleted there is no way to download the same episode again and that's kind of bummer.

Is this the normal behavior? Coudl it be changed, so deleting a file/episode, it doesn't get deleted from the list?

Thanks,

**Comment by randomnicode** [comment]
Hmm, I believe this is the key function:
https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java#L717-L746

Haven't dug in but there does seem to be a logical delete option there that doesn't delete from the db. In fact that seems to be the default delete. Want to dig in more?

**Comment by llucps** [comment]
Hi @randomnicode,

Thanks to point me out on the right direction, although I can read and understand a bit of code unfortunately I'm not a developer, so what I've done might not be the best option :)

Since what I want is too remove the file and keep the episode reference in the podcast list, I changed this piece of code:

`
if (logicalDelete) {
            episode.setStatus(PodcastStatus.DELETED);
            episode.setErrorMessage(null);
            podcastDao.updateEpisode(episode);
        } else {
            podcastDao.deleteEpisode(episodeId);
        }
`

for

`
if (logicalDelete) {
            episode.setStatus(PodcastStatus.SKIPPED);
            episode.setErrorMessage(null);
            podcastDao.updateEpisode(episode);
        } else {
            episode.setStatus(PodcastStatus.SKIPPED);
        }
`
and it works, I basically set the status to SKIPPED which is the one be default.

Does it make sense?

Thanks,

**Comment by randomnicode** [comment]
If it works for you locally that's fine then.

However, my recommendation for you would actually be to figure out why a DELETED episode isn't showing up when you retrieve the list. Something is likely filtering DELETED episodes while showing SKIPPED since either way the episode stays in the DB.

**Comment by llucps** [comment]
Yes you're right.. it doesn't do anything.. so for the moment so changing PodcastStatus.DELETED for PodcastStatus.SKIPPED it does the job.. although it's not the solution as you said.

I'm trying to understand what the deleted episodes are not shown since they are still referenced in the DB after the file is deleted.

It's not easy to me to figure out the code... but in the function getEpisode seems to be the one the retrieve the episodes and list them. The function has to parameters, the episodeID (int) and includeDeleted (boolean), and inside there is conditional where says if the episode status is deleted and is different than includeDeleted then return nothin (null).

`public PodcastEpisode getEpisode(int episodeId, boolean includeDeleted) {
        PodcastEpisode episode = podcastDao.getEpisode(episodeId);
        if (episode == null) {
            return null;
        }
        if (episode.getStatus() == PodcastStatus.DELETED && !includeDeleted) {
            return null;
        }
        addMediaFileIdToEpisodes(Arrays.asList(episode));
        return episode;
    }`

so at the first glance I thought to add a true as second parameter in:

`PodcastEpisode episode = podcastDao.getEpisode(episodeId, true);`

but I get this error during compilation:

`/var/lib/tomcat8/airsonic-dev/airsonic/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java:[241,44] method getEpisode in class org.airsonic.player.dao.PodcastDao cannot be applied to given types;
[ERROR]   required: int
[ERROR]   found: int,boolean
[ERROR]   reason: actual and formal argument lists differ in length`

I think I'm on the right path... but it's not easy for me to grasp all the code :)

Any ideas?

Thanks

**Comment by randomnicode** [comment]
Right, so two things, first some general concepts, and then where I think your issue is:

1.
I think you're confusing two separate methods named similarly in two separate classes. Sorry this might be a little confusing. It's usually helpful to beginners to see it in an IDE so they can "visualize" it, and see and mentally map the references. Plus be familiar with some convention and abstraction (sorry again). If you click on a particular method call, and explore the options, most IDEs will let you go to its definition that will clue you in to what method is being called despite similar names.

The two operative classes are:
- PodcastDao: This talks to the database (DAO in its name is an acronym for Data Access Object. By convention anything ending with that name will usually talk to the DB). It talks SQL, the language of the database. This is this class: https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java
- PodcastService: This serves as the intermediate layer between the incoming request from the user and where the data actually resides (in the DB). It talks to the DB through the Dao mentioned (and not directly), and "orchestrates" most of what's going on and how it's going to happen. This is this class: https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java

This method: 
`public PodcastEpisode getEpisode(int episodeId, boolean includeDeleted)` 

is calling _podcastDao_.getEpisode(episodeId). It's not calling itself again (that would be recursion).

`podcastDao.getEpisode()` is here: 
https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java#L150

And you can see it only takes in a single argument. Thus your compilation error.

In short, the service (a) retrieves the episode by telling the dao to retrieve it from the DB, (b) then filters it depending on whether its status is DELETED or not. PodcastDao retrieves it from the DB with a single argument, the episode id, it doesn't take in another argument because it doesn't need it, it doesn't filter or anything.

2.
I suspect this isn't your issue though. I don't know what view you're looking at (and I have no idea what podcast views there are), but I suspect you're looking at a list of episodes for a given channel.
In that case you should likely be looking at `getEpisodes(int channelId)`
which calls the PodcastDao here: https://github.com/airsonic/airsonic/blob/master/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java#L125

and you can see how it filters out from the DB in the query `and status != ? ... PodcastStatus.DELETED.name()`

Now you can remove that part from the sql query and the DELETED argument, and your db should start returning all the records for a channel.

No idea how to redownload stuff or if it throws an exception in the UI.
